Mathlib/Data/Matrix/Composition.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2024 Yunzhou Xie. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Yunzhou Xie
+-/
+
+import Mathlib.Data.Matrix.Basic
+
+/-!
+## Main results
+# Composition of matrices
+This file shows that Mₙ(Mₘ(R)) ≃ Mₙₘ(R), Mₙ(Mₘ(R)) ≃ Mₘ(Mₙ(R)), Mn(Rᵒᵖ) ≃ₐ[K] Mₙ(R)ᵒᵖ
+and also different levels of equivalence when R is an AddCommMonoid,
+Semiring, and Algebra over a CommSemiring K.
+-/
+
+namespace Matrix
+
+variable  (I J K L R : Type*)
+
+/-- Mₙ(Mₘ(R)) ≃ Mₙₘ(R) -/
+@[simps]
+def comp : Matrix I J (Matrix K L R) ≃ Matrix (I × K) (J × L) R where
+  toFun m ik jl := m ik.1 jl.1 ik.2 jl.2
+  invFun n i j k l := n (i, k) (j, l)
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- Mₙ(Mₘ(R)) ≃ Mₘ(Mₙ(R)) -/
+@[simps]
+def swap : Matrix (I × J) (K × L) R ≃ Matrix (J × I) (L × K) R where
+  toFun m ji kl := m (ji.2, ji.1) (kl.2, kl.1)
+  invFun n ij kl := n (ij.2, ij.1) (kl.2, kl.1)
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+section AddCommMonoid
+
+variable [AddCommMonoid R]
+
+/-- Mₙ(Mₘ(R)) ≃+ Mₙₘ(R) -/
+@[simps!]
+def compAddHom : Matrix I J (Matrix K L R) ≃+ Matrix (I × K) (J × L) R :=
+{ Matrix.comp I J K L R with
+  map_add' := fun _ _ ↦ rfl
+}
+
+/-- Mₙ(Mₘ(R)) ≃+ Mₘ(Mₙ(R)) -/
+@[simps!]
+def swapAddHom : Matrix (I × J) (K × L) R ≃+ Matrix (J × I) (L × K) R :=
+{ Matrix.swap I J K L R with
+  map_add' := fun _ _ ↦ rfl
+}
+
+end AddCommMonoid
+
+section Semiring
+
+variable [Semiring R] [Fintype I] [Fintype J] [DecidableEq I] [DecidableEq J]
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₙₘ(R) -/
+@[simps!]
+def compRingHom : Matrix I I (Matrix J J R) ≃+* Matrix (I × J) (I × J) R :=
+{ Matrix.compAddHom I I J J R with
+  map_mul' := fun _ _ ↦ by
+    ext _ _
+    refine (Matrix.sum_apply _ _ _ _).trans $ Eq.symm Fintype.sum_prod_type
+}
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₘ(Mₙ(R)) -/
+@[simps!]
+def swapRingHom : Matrix (I × J) (I × J) R ≃+* Matrix (J × I) (J × I) R :=
+{ Matrix.swapAddHom I J I J R with
+  map_mul' := fun _ _ ↦ by
+    ext _ _
+    simp only [swapAddHom, swap, AddEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe, EquivLike.coe_coe,
+      AddEquiv.coe_mk, Equiv.coe_fn_mk, mul_apply, Fintype.sum_prod_type]
+    exact Finset.sum_comm
+}
+
+end Semiring
+
+section LinearMap
+
+variable (K : Type*) [CommSemiring K] [AddCommMonoid R] [Module K R]
+
+/-- Mₙ(Mₘ(R)) ≃ₗ[K] Mₙₘ(R) -/
+@[simps!]
+def compLinearMap : Matrix I J (Matrix K L R) ≃ₗ[K] Matrix (I × K) (J × L) R :=
```

Comment:
This should be compLinearEquiv and similarly for algEquiv

Reply:
oh okay